### PR TITLE
166214098 verify email address

### PR DESCRIPTION
**Pull request by JuwonAbiola**
#### What does this PR do?
verify email address

#### Description of Task to be completed?
-user should be able to verify their email address

#### How should this be manually tested?
clone or download the project
checkout to ft-verify-email-address-166214098
git pull to get the latest update
npm i to install dependencies
npm start to start the server
signup using postman
check your mail  
click on the verify link

#### Any background context you want to provide?
N/A

#### What are the relevant pivotal tracker stories?
[166214098](https://www.pivotaltracker.com/n/projects/2350588/stories/166214098)

Screenshots (if appropriate)

image
image
image

Questions:
N/A

### CODE REVIEW

**Review by Lundii — COMMENTED** [review]
I followed the instructions on your PR but didn't get any mail
image

image

File: server/helpers/createToken.js [hunk 1]
```diff
@@ -3,9 +3,11 @@ import dotenv from 'dotenv';
 
 dotenv.config();
 
-const createToken = (payload, expiresIn) => {
+export const createToken = (payload, expiresIn) => {
   if (!expiresIn) return jwt.sign(payload, process.env.JWT_KEY);
```
Comment (henryade) [inline server/helpers/createToken.js]: use export default

File: server/helpers/index.js [hunk 2]
```diff
@@ -1,9 +1,12 @@
-import createToken from './createToken';
+import { createToken, verifyToken } from './createToken';
 import responseMessage from './responseMessage';
```
Comment (henryade) [inline server/helpers/index.js]: destructure after the import not in the import statement

File: server/index.js [hunk 3]
```diff
@@ -9,7 +9,7 @@ import routes from './routes';
 
 config();
 
-const { PORT } = process.env; // setup port to be used
+export const { PORT } = process.env; // setup port to be used
```
Comment (henryade) [inline server/index.js]: why are you exporting port from here?

File: test/server/routes/user.test.js [hunk 4]
```diff
@@ -13,9 +13,12 @@ const User = {
   password: 'joejoe',
   avatarUrl: 'https://gravatar.com/avatar/aed61baf1e9256ed7d70e2cbbfcba9aa?s=400&d=robohash&r=x',
 };
+
+let token = '';
 const userSignup = '/signup';
 
-// Test Signing up a user
+console.log("hello", token)
```
Comment (henryade) [inline test/server/routes/user.test.js]: Forgot something?

File: server/routes/user.js [hunk 5]
```diff
@@ -3,8 +3,9 @@ import UserController from '../controllers/userController';
 
 const user = express.Router();
 
-const { createUser } = UserController;
+const { createUser, verifyUser } = UserController;
 
 user.post('/signup', createUser);
+user.get('/verify/:token', verifyUser);
```
Comment (KvNGCzA) [inline server/routes/user.js]: use the verifyToken middleware on this route

**Review by KvNGCzA — CHANGES_REQUESTED** [review]

File: server/controllers/userController.js [hunk 6]
```diff
@@ -25,13 +25,27 @@ class UserController {
       const newUser = await User.create({ firstName, lastName, email, password: bcrypt.hashSync(password, 10), avatarUrl, verified: false })
       if (newUser) {
         const { id } = newUser.dataValues
-        const token = createToken({ id }, '24h');
+        const token = createToken({ id, email }, '24h');
```
Comment (KvNGCzA) [inline server/controllers/userController.js]: why are you adding email here? only `id` should be here

File: server/controllers/userController.js [hunk 7]
```diff
@@ -25,13 +25,27 @@ class UserController {
...
       }
     } catch (error) {
       return responseMessage(response, 500, { message: error.message });
     }
   }
+
+  static async verifyUser(request, response) {
+    const { email } = request.userData;
+    try {
+      const alreadyVerified = await User.findOne({ where: { verified: true } })
```
Comment (KvNGCzA) [inline server/controllers/userController.js]: This should have been removed by now

File: server/helpers/createToken.js [hunk 8]
```diff
@@ -3,9 +3,11 @@ import dotenv from 'dotenv';
...
   return jwt.sign(payload, process.env.JWT_KEY, { expiresIn });
 };
 
-export default createToken;
+export const verifyToken = (payload) => {
```
Comment (KvNGCzA) [inline server/helpers/createToken.js]: this function is not in use

File: server/index.js [hunk 9]
```diff
@@ -10,7 +10,7 @@ import routes from './routes';
...
```
Comment (KvNGCzA) [inline server/index.js]: why are we doing this?

File: test/server/routes/auth.test.js [hunk 10]
```diff
@@ -108,99 +110,31 @@ describe('AUTH', () => {
           done();
         });
     });
-  });
```
Comment (KvNGCzA) [inline test/server/routes/auth.test.js:207]: What happened to all other tests in this file?

**Review by KvNGCzA — CHANGES_REQUESTED** [review]
Please attend to all feedbacks

File: test/server/routes/auth.test.js (hunk 10 above)
Comment (JuwonAbiola) [reply test/server/routes/auth.test.js:207]: they are there

File: server/index.js (hunk 9 above)
Comment (JuwonAbiola) [reply server/index.js]: fixed

File: server/controllers/userController.js (hunk 6 above)
Comment (JuwonAbiola) [reply server/controllers/userController.js]: fixed

File: server/controllers/userController.js (hunk 7 above)
Comment (JuwonAbiola) [reply server/controllers/userController.js]: done

File: server/helpers/createToken.js (hunk 8 above)
Comment (JuwonAbiola) [reply server/helpers/createToken.js]: removed

**Review by KvNGCzA — CHANGES_REQUESTED** [review]
Write tests to cover all the edge cases you have written

**Review by Lundii — CHANGES_REQUESTED** [review]
You need to rebase first. The develop is ahead of the PR, probably that's why the test coverage reduced significantly.

File: server/middlewares/verifyToken.js [hunk 11]
```diff
@@ -10,18 +10,18 @@ export default (request, response, next) => {
   const token = request.headers.authorization || request.query.token;
 
   if (token) {
-    jwt.verify(token, process.env.JWT_KEY, (error, decoded) => {
+    jwt.verify(token, process.env.JWT_KEY, async(error, decoded) => {
```
Comment (allebd) [inline server/middlewares/verifyToken.js:13]: Missing space before function parentheses

File: test/server/routes/auth.test.js [hunk 12]
```diff
@@ -202,5 +204,31 @@ describe('AUTH', () => {
           done();
         });
     });
+
+    it('It Should verify user email address', (done) => {
+      chai.request(server)
+        .get(`${BASE_URL}/verify?token=${token}`)
+        .send(User)
+        .end((error, response) => {
+          console.log(response);
```
Comment (allebd) [inline test/server/routes/auth.test.js]: I do not think you need console.log here

**Review by Lundii — DISMISSED** [review]

**Review by henryade — DISMISSED** [review]

**Review by KvNGCzA — APPROVED** [review]
